app/src/main/java/de/westnordost/streetcomplete/quests/recycling_material/AddRecyclingContainerMaterials.kt

```diff
@@ -51,88 +51,72 @@ class AddRecyclingContainerMaterials : OsmElementQuestType<RecyclingContainerMat
     override fun getHighlightedElements(element: Element, getMapData: () -> MapDataWithGeometry) =
         getMapData().filter("nodes with amenity = recycling")
 
-    override fun applyAnswerTo(answer: RecyclingContainerMaterialsAnswer, changes: StringMapChangesBuilder) {
+    override fun applyAnswerTo(answer: RecyclingContainerMaterialsAnswer, tags: StringMapChangesBuilder) {
         if (answer is RecyclingMaterials) {
-            applyRecyclingMaterialsAnswer(answer.materials, changes)
+            applyRecyclingMaterialsAnswer(answer.materials, tags)
         } else if(answer is IsWasteContainer) {
-            applyWasteContainerAnswer(changes)
+            applyWasteContainerAnswer(tags)
         }
     }
 
-    private fun applyRecyclingMaterialsAnswer(materials: List<RecyclingMaterial>, changes: StringMapChangesBuilder) {
+    private fun applyRecyclingMaterialsAnswer(materials: List<RecyclingMaterial>, tags: StringMapChangesBuilder) {
+        // first clear recycling:* taggings previously "yes"
+        val previousMaterials = tags.entries.filter { (key, value) ->
+            key.startsWith("recycling:") && value == "yes"
+        }.map { it.key }
+        for (material in previousMaterials) {
+            tags.remove(material)
+        }
```

Comment:
I really like the simplification in this function!

For this part, this would be a bit shorter and saves two loop iterations:

```suggestion
        for ((key, value) in tags.entries) {
            if (key.startsWith("recycling:") && value == "yes") {
                tags.remove(material)
            }
        }
```